### ISSUE TITLE
Remove unused usings on format

### ISSUE THREAD

**Issue by lsegal** [OmniSharp/omnisharp-vscode#5261]
## Is your feature request related to a problem? Please describe.

tl;dr: OmniSharp should allow configuring the formatter (omnisharp.json) to remove unused imports.

#1245 shows that it is possible to "remove unused usings" in a document using Omnisharp. However, this feature cannot be used during a "Format" operation (as per comment https://github.com/OmniSharp/omnisharp-vscode/issues/1245#issuecomment-763723115). Only "organizing" imports is possible via the omnisharp.json configuration (#1686).

## Describe the solution you would like

OmniSharp should allow configuring the formatter (omnisharp.json) to remove unused imports.

## Applicable Scenarios

* Ctrl+P >Format Document
* Format on Save

**Comment by Meligy** [comment]
I wonder if there's any documentation on what fixes are considered in formatting and what are not?

**Comment by devonpmack** [comment]
Add this to your settings.json and it should do it.

```
  "[csharp]": {
    "editor.defaultFormatter": "ms-dotnettools.csharp",
    "editor.codeActionsOnSave": ["source.organizeImports", "source.fixAll"]
  },
```

**Comment by Tantol** [comment]
> Add this to your settings.json and it should do it.
> 
> ```
>   "[csharp]": {
>     "editor.defaultFormatter": "ms-dotnettools.csharp",
>     "editor.codeActionsOnSave": ["source.organizeImports", "source.fixAll"]
>   },
> ```

Not working for me. Unused usings still exists after save.

Any other solutions or plans to implement this feature in to omnisharp?

**Comment by Meligy** [comment]
For this to work, I read somewhere that you need to turn on XML comments (sorry, I don't remember where):

So, in my project (in my `Directory.Build.props` file really, which applies to all projects(, I also had:

```xml
  <PropertyGroup>
    <!-- Needed to help with 'usings' removal, as some are not removed in case they are in XML comments -->
    <GenerateDocumentationFiles>true</GenerateDocumentationFiles>
  </PropertyGroup>
```

Then I needed to have an `.editorconfig` file, which had the following rules for C# files, among other rules that I had anyway:

```yaml
# unused directives
dotnet_diagnostic.IDE0005.severity = none
dotnet_diagnostic.CS8019.severity = silent
# Disable: Missing XML comment for publicly visible type or member
# We only turned on XML comments to help with detecting unused usings
dotnet_diagnostic.CS1591.severity = none
```

A few other relevant but not necessary `.editorconfig` rules:

```yaml
# File-scoped namespace declarations
dotnet_diagnostic.IDE0161.severity = refactoring
csharp_style_namespace_declarations = file_scoped
# Sort using and Import directives with System.* appearing first
dotnet_sort_system_directives_first = true
dotnet_separate_import_directive_groups = false
dotnet_style_namespace_match_folder = true
dotnet_diagnostic.IDE0130.severity = warning
```

**Comment by Sweepertank** [comment]
> Add this to your settings.json and it should do it.
> 
> ```
>   "[csharp]": {
>     "editor.defaultFormatter": "ms-dotnettools.csharp",
>     "editor.codeActionsOnSave": ["source.organizeImports", "source.fixAll"]
>   },
> ```

This works for me, but `source.fixAll` seems to be doing more than just removing unused usings - for instance, it also causes stub implementations of unimplemented interface members to be generated. It would be great if there were an option _just_ to remove unused usings on save without enabling any other on-save actions.

**Comment by Meligy** [comment]
> It would be great if there were an option _just_ to remove unused usings on save without enabling any other on-save actions.

You are expected to control which fixes are enabled and which are not using the editorconfig file.

Docs https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-options?source=recommendations

**Comment by terdong** [comment]
To work, check this option

![image](https://github.com/dotnet/vscode-csharp/assets/7655772/04ef7b41-97e7-4094-ac9a-c98b3aa59de0)

**Comment by IanIsFluent** [comment]
> To work, check this option
> 
> ![image](https://user-images.githubusercontent.com/7655772/268669948-04ef7b41-97e7-4094-ac9a-c98b3aa59de0.png)

Not working for me. 👎

**Comment by JoeRobich** [comment]
@IanIsFluent If you are using OmniSharp, you would need to additionally enable `omnisharp.organizeImportsOnFormat`

**Comment by terdong** [comment]
> > To work, check this option
> > ![image](https://user-images.githubusercontent.com/7655772/268669948-04ef7b41-97e7-4094-ac9a-c98b3aa59de0.png)
> 
> Not working for me. 👎
I experienced other side effects even though the Remove unused import in omnisharp worked.
That's way I stopped using csharp in vscode and went back to visual studio. I think omnisharp's latest support is out of date. I'm sorry I can not help.

**Comment by mcuthbe** [comment]
@terdong's fix works for me, but does anyone know why it doesn't work with the newer LSP powered version of the extension? Is it a roslyn thing or due to the extension itself? Struggling to find any information about it

**Comment by HSPiira** [comment]
Any one still struggling with this, do this:

Press Ctrl + . (Control key + period key) when the cursor is inside the file. 
This opens up the Quick Actions menu, where you can choose "Remove Unnecessary Usings"

**Comment by JeremyCaron** [comment]
I maintain an extension called [C# Organize Usings](https://marketplace.visualstudio.com/items?itemName=jeremycaron.csharp-organize-usings) ([repo](https://github.com/JeremyCaron/vscode-csharp-organize-usings)) that can also handle this, along with sorting & spacing as well.  It's compatible with both Omnisharp and Roslyn.

**Comment by sunefred** [comment]
@JeremyCaron Of all the suggested solutions in this thread, on Google and suggested by LLMs, this is the only one that worked. Why C# Dev Kit can not do this out of the box is surprising, but I am happy I found this extension! 🙏

**Comment by AMArostegui** [comment]
The LLM suggested commands both in VSCode palette and mouse context menu, but I couldn't find any.

In my case, what works is to place the cursor on an unused using (grayed-out so these are easy to notice) and type Ctrl+. Then in the lightbulb menu I get "Remove all unnecesary usings" command

I'm using C# Dev Kit 1.70.3
